### ISSUE TITLE
MainTemplate plugin hooks stopped working in webpack 5 even though a compat layer is supposed to exist

### ISSUE THREAD

**Issue by bdwain** [webpack/webpack#13294]
# Bug report

**What is the current behavior?**
The `mainTemplate` hooks that I used in a plugin in webpack 4 seem to no longer get called on the webpack runtime code in webpack 5. 

**If the current behavior is a bug, please provide the steps to reproduce.**

```js
function MyPlugin() {}
MyPlugin.prototype.apply = function (compiler) {
  compiler.hooks.compilation.tap(
    'MyPlugin',
    function (compilation) {
      compilation.mainTemplate.hooks.requireExtensions.tap(
        'MyPlugin',
        (source) =>
          source.replace(
            'Object.defineProperty(exports, key, { enumerable: true, get: definition[key] });',
             myNewCode
          )
      );
    }
  );
};
```

From what I can tell, the callback does get called still, but source is always empty, so my plugin doesn't actually do anything anymore. I've also tried different hooks on mainTemplate, but none of them seemed to work.

**What is the expected behavior?**
According to the [migration guide](https://webpack.js.org/blog/2020-10-10-webpack-5-release/#mainchunkmoduletemplate-deprecation), while we should be moving to the `JavascriptModulesPlugin`, 

```
There is a compat-layer, so Main/Chunk/ModuleTemplate still exist, but only delegate tap calls to the new hook locations.
```

The new `JavascriptModulesPlugin` is not documented at all, so it's hard to follow the migration instructions. But I would have assumed the old hooks would still work anyway while the compat layer exists. I tried messing with the `renderMain` hook in the JavascriptModulesPlugin, but it did not seem to be what I was looking for. 

**Other relevant information:**
webpack version:  5.36.2
Node.js version:  14.16.0
Operating System:  OS X Catalina
Additional tools:

**Comment by alexander-akait** [comment]
We need update docs
